[18.0][MIG] shopfloor_packing -> shopfloor_cluster_picking_repack

sebalix opened the pull request:
Checked with @jbaudoux , we removed the test `test_put_in_pack_partial_qty` as it's not something supported by Odoo 18.0 in std.
Also, the module has been renamed to `shopfloor_cluster_picking_repack` as it is an extension of the `cluster_picking` scenario.

Frontend/mobile modules will be ported through other PRs.

Depends on:
- https://github.com/OCA/delivery-carrier/pull/973
- https://github.com/OCA/delivery-carrier/pull/1043
- https://github.com/OCA/stock-logistics-tracking/pull/51
- https://github.com/OCA/stock-logistics-putaway/pull/11
- https://github.com/OCA/stock-logistics-shopfloor/pull/45

sbejaoui committed:
[16.0][ADD] - shopfloor_packing

sbejaoui committed:
[IMP] description

sbejaoui committed:
[IMP] add failing test for packing partial qty

If a partially available product becomes completely available and the reserved
quantity increases after the "put in pack" action, a new move line will be
created with the done quantity as a copy of the line that will be packed.
The original line will keep the result package, which is the package scanned
by the operator, and the remaining quantity to do.

Since the package is an internal package, the line with the remaining quantity
to do will be proposed in the packing process and considered as picked. If the
operator don't pay attention, and confirm the packaging, both lines will be set
to done.

sbejaoui committed:
[FIX] packing partial qty

rousseldenis committed:
[IMP] shopflor/mobile/_packing: Allows to select package types for packing

rousseldenis committed:
[IMP] shopfloor_packing: Move data retrieval to data action + enable tests

rousseldenis committed:
[IMP] shopfloor_packing: Use better message

rousseldenis committed:
[IMP] shopfloor_packing: Use appropriate name for packing action + improve tests

rousseldenis committed:
[IMP] shopfloor_packing: Several cosmetics + update list_delivery_packaging endpoint + add tests

rousseldenis committed:
[IMP] shopfloor_packing: Update docstring for package types

rousseldenis committed:
[IMP] shopfloor_packing: Improve helps + add contributors

rousseldenis committed:
[IMP] shopfloor_packing: Move BE translation to generic fr one

rousseldenis committed:
[IMP] shopfloor_*_packing: Improve packing flow

lmignon committed:
[IMP] shopfloor_packing: Avoid concurrent update

Simlify the content of the message returned when packing a picking or selecting a package. The most important change it to not include into the product's information the qty_available field. The computation of this useless info in such context could lead to concurrent update since it requires to select all the quants into the stock while a lot of operations on the quants are done in parallel.

lmignon committed:
[FIX] shopfloor_packing: Ensure number of parcels is preserved

if a number of parcels is specified and a package type with a different default value is used, we must ensure that at the end of the process the encoded number of parcels is the one specified into the UI

sbejaoui committed:
[IMP] shopfloor_packing: compute result package weight through move lines

there are two methods to compute the result package estimated weight: using move lines or using quants

using move lines is more efficient and avoids concurrent access on stock quants.
to enable this approach, the picking_id must be set in the context

sbejaoui committed:
[IMP] shopfloor_packing: skip package quantity computation in packing scenario

sebalix committed:
[MOV] Move shopfloor_packing to shopfloor_cluster_picking_repack

sebalix committed:
[IMP] shopfloor_cluster_picking_repack: pre-commit auto fixes

sebalix committed:
[MIG] shopfloor_cluster_picking_repack: Migration to 18.0

jbaudoux commented:
cc @rousseldenis @sbejaoui

jbaudoux commented on shopfloor_cluster_picking_repack/tests/test_cluster_picking_pack_picking.py:
```diff
@@ -335,78 +335,6 @@ def test_response_for_scan_destination(self):
 
-    def test_put_in_pack_partial_qty(self):
-        """
-        Special case.
```
Note for reviewers: removed as in v18, you cannot assign a picked move, so you never have a new move line automaticaly created

BhaveshHeliconia reviewed (approved):
LGTM!

sebalix commented on shopfloor_cluster_picking_repack/models/stock_quant_package.py:
```diff
@@ -0,0 +1,13 @@
+class StockQuantPackage(models.Model):
+    _inherit = "stock.quant.package"
+
+    def _sync_package_type_from_single_product(self):
```
The signature of that one changed:
```suggestion
    def _sync_package_type_from_single_product(self, product, quantity):
```

jbaudoux commented on shopfloor_cluster_picking_repack/services/packing.py:
```diff
@@ -0,0 +1,77 @@
+# Copyright 2020-2021 Camptocamp SA (http://www.camptocamp.com)
+# Copyright 2020-2021 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
```
Remove end date
```suggestion
# Copyright 2020 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
```

jbaudoux commented on shopfloor_cluster_picking_repack/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    "name": "Shopfloor Packing",
+    "version": "18.0.1.0.0",
+    "summary": """ Manage Packing into cluster picking""",
+    "author": "ACSONE SA/NV,Odoo Community Association (OCA)",
```
I should have been listed here also 
```suggestion
    "author": "ACSONE SA/NV,BCIM,Odoo Community Association (OCA)",
```

jbaudoux commented on shopfloor_cluster_picking_repack/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Shopfloor Packing",
```
```suggestion
    "name": "Shopfloor Cluster Picking Repack",
```

jbaudoux commented on shopfloor_cluster_picking_repack/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "Shopfloor Packing",
+    "version": "18.0.1.0.0",
+    "summary": """ Manage Packing into cluster picking""",
```
```suggestion
    "summary": """Build delivery packages at the end of the cluster picking""",
```

jbaudoux committed:
[FIX] shopfloor_cluster_picking_repack: fix migration for package type

jbaudoux commented:
Added https://github.com/OCA/stock-logistics-shopfloor/pull/45 as dependency as pushed required related fixes

TDu commented on shopfloor_cluster_picking_repack/services/cluster_picking.py:
```diff
@@ -0,0 +1,605 @@
+        res["data"]["schema"]["pack_picking_scan_pack"] = self._schema_pack_picking
+        res["data"]["schema"]["select_package"] = self._schema_select_package
+        return res
+
```
Without this, when hitting a zero stock on scanning the destination pack,  the data dictionary for the next_ state `zero_check` is empty !
```
def is_zero(self) -> dict:              
    res = super().is_zero()
    res["data"]["schema"]["select_package"] = self._schema_select_package

```

simahawk committed:
shopfloor_cluster_picking_repack: fix qty on internal bin error

Selected qty must be kept always in case of errors.

simahawk commented:
@jbaudoux `TestClusterPickingPrepareUnload.test_put_in_pack_cant_set_correct_package_type` is failing

TDu committed:
shopfloor_cluster_picking_repack: fix 1 unit test

The auto assign packaging from the module
`stock_quant_package_product_packaging`
will as a last option assign the default package type of the product
on the result package if there is none.
Making the check after the action done fail.

simahawk committed:
shopfloor_cluster_picking_repack: update contribs

simahawk commented:
/ocabot merge nobump
